### PR TITLE
feat/tdd

### PR DESCRIPTION
**Pull request by Lainera**
Add tdd
Add Vagrantfile
Add playbook

Signed-off-by: Lainera <fistlain@hotmail.com>

### CODE REVIEW

File: Readme.md [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+  string executable = 1;
+  optional CpuControl cpu_control = 2;
+  optional MemControl mem_control = 3;
+  optional IoControl io_control = 4;
+  repeated string args = 5;
+}
+
+message Ack {}
+
+message JobId { bytes jobid = 1; }
```
Comment (r0mant) [inline Readme.md:150]: What is used as a job ID?

File: Readme.md [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+- `cpu.weight`: provide value in [1, 100000] to set weight for spawned process (default 100), relative to other client processes.
+- `memory.high`: provide value in bytes to cap memory, process using more than this value will be throttled.
+- `memory.max`: provide value in bytes to hard-cap memory, process using more than this value will be OomKilled. 
+- `io.max.rbps` & `io.max.wbps`: provide value to cap io in bytes per second. 
+
+### Running as root
+
+Adding process to a cgroup requires eid of the process performing the move to be the owner of common ancestor of source and destination cgroups. In other words, for child C of process P belonging to non-root user U, to be moved in a cgroup owned by U, P itself needs to be in a cgroup that is direct ancestor of target cgroup and owned by U.   
+This could be achieved by setting up appropriate permissions beforehand as privileged user and executing binary as U.
+For this implementation it is a no-goal and process is ran as root.
```
Comment (r0mant) [inline Readme.md:29]: Could you expand on your process execution model? It's not very clear from this description how the server will launch the jobs and place them into cgroups.

File: Readme.md [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+
+Notes:
+- Considered keeping logs in-memory, but I believe this approach has couple of drawbacks. Reads and writes need to be synchronized somehow, so our options are either introducing Mutex, which may become fairly hot (depends on the volume of logs), or making Job own the output of the child process instead of streaming to file. This consequently means that all job output is now in main process memory, regardless of whether it will ever be read or not.
```
Comment (r0mant) [inline Readme.md]: How are clients going to get the new output produced by the job in real-time?

**Review by xacrimon — APPROVED** [review]

**Comment by xacrimon** [comment]
LGTM once Roman's comments are sorted.

File: Readme.md (hunk 2 above)
Comment (Lainera) [reply Readme.md:29]: Added **Flow** section below, pls lmk if I'm missing something and/or any further clarifications are needed. 
Do mermaid diagrams open for you? If for some reason they do not - code could be pasted @[mermaid live editor](https://mermaid.live/)

File: Readme.md (hunk 1 above)
Comment (Lainera) [reply Readme.md:150]: Intention is to use [uuid](https://docs.rs/uuid/latest/uuid/), added comment.

File: Readme.md (hunk 3 above)
Comment (Lainera) [reply Readme.md]: Good callout, added a note on `inotify`

File: Readme.md [hunk 4]
```diff
@@ -0,0 +1,375 @@
...
+
+Notes:
+- Considered keeping logs in-memory, but I believe this approach has couple of drawbacks. Reads and writes need to be synchronized somehow, so our options are either introducing Mutex, which may become fairly hot (depends on the volume of logs), or making Job own the output of the child process instead of streaming to file. This consequently means that all job output is now in main process memory, regardless of whether it will ever be read or not. 
+- Assuming job is still running, only logs that were generated at the moment of request would be sent back. To address that, controller could leverage [inotify](https://man7.org/linux/man-pages/man7/inotify.7.html) (intention is to use [inotify-rs](https://docs.rs/inotify/latest/inotify/)). Given log streamer task read EOF from open file, instead of exiting it `select!`'s on one of two events - either job done, in which case no new logs were generated or `IN_MODIFY` event, in which case file is opened again, `seek`'ed to wherever cursor was left and streaming to client resumes.
```
Comment (r0mant) [inline Readme.md:279]: > only logs that were generated at the moment of request would be sent back

Just wanted to clarify what this means exactly - will the client still receive all logs produced by the job from the beginning? That is part of the challenge spec requirements.

Comment (Lainera) [reply Readme.md:279]: Yes this is correct, each output request gets logs from the beginning of the log file and continues receiving logs as they come in.

**Review by r0mant — APPROVED** [review]

**Pull request merged**